airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/discover/DiscoverOperation.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2025 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.discover
+
+import io.airbyte.cdk.Operation
+import io.airbyte.cdk.command.ConfigurationSpecification
+import io.airbyte.cdk.command.ConfigurationSpecificationSupplier
+import io.airbyte.cdk.load.command.Append
+import io.airbyte.cdk.load.command.Dedupe
+import io.airbyte.cdk.load.command.DestinationConfiguration
+import io.airbyte.cdk.load.command.DestinationConfigurationFactory
+import io.airbyte.cdk.load.command.DestinationDiscoverCatalog
+import io.airbyte.cdk.load.command.DestinationOperation
+import io.airbyte.cdk.load.command.ImportType
+import io.airbyte.cdk.load.command.SoftDelete
+import io.airbyte.cdk.load.command.Update
+import io.airbyte.cdk.load.data.json.AirbyteTypeToJsonSchema
+import io.airbyte.cdk.output.ExceptionHandler
+import io.airbyte.cdk.output.OutputConsumer
+import io.airbyte.protocol.models.v0.DestinationCatalog as ProtocolDestinationCatalog
+import io.airbyte.protocol.models.v0.DestinationOperation as ProtocolDestinationOperation
+import io.airbyte.protocol.models.v0.DestinationSyncMode as ProtocolDestinationSyncMode
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Requires
+import jakarta.inject.Singleton
+
+private val logger = KotlinLogging.logger {}
+
+@Singleton
+@Requires(property = Operation.PROPERTY, value = "discover")
+@Requires(env = ["destination"])
+class DiscoverOperation<T : ConfigurationSpecification, C : DestinationConfiguration>(
+    val configJsonObjectSupplier: ConfigurationSpecificationSupplier<T>,
+    val configFactory: DestinationConfigurationFactory<T, C>,
+    val destinationDiscoverer: DestinationDiscoverer<C>,
+    private val exceptionHandler: ExceptionHandler,
+    private val outputConsumer: OutputConsumer,
+) : Operation {
+
+    override fun execute() {
+        val pojo =
+            try {
+                configJsonObjectSupplier.get()
+            } catch (e: Exception) {
+                handleException(e)
+                return
+            }
+        val config =
+            try {
+                configFactory.make(pojo)
+            } catch (e: Exception) {
+                handleException(e)
+                return
+            }
+        try {
+            val destinationCatalog = destinationDiscoverer.discover(config)
+            outputConsumer.accept(destinationCatalog.toProtocol())
+        } catch (t: Throwable) {
+            logger.warn(t) { "Caught throwable during DISCOVER" }
+            handleException(t)
+        } finally {
+            destinationDiscoverer.cleanup(config)
```

Comment:
curious: do we expect to need cleanup? this cleanup step is useful in `check`, where we might be creating temp tables/etc during the `check` operation

but I would have guessed that discover is typically a read-only thing? so we shouldn't need to clean up after ourselves

Reply:
I don't have anything in mind at the moment. I replicated what we did for check. I am assuming it doesn't hurt if that's a noop and it might be convenient if there's anything to clean up.